### PR TITLE
Add event API to have a first view of notable event (Thread create and terminate, Module load and unload)

### PR DESCRIPTION
**Pull request by citronneur**
Hi @commial !

Here a PR that add the event API of TTD on the engine object. This API is very useful to create a firts macro timeline to point directly at an interesting position, like the load of net module.

Have a nice day,

Sylvain

### CODE REVIEW

File: TTD/TTD.hpp [hunk 1]
```diff
@@ -7,50 +9,40 @@ namespace TTD {
...
-		unsigned __int64 r13;
-		unsigned __int64 r14;
-		unsigned __int64 r15;
-		unsigned __int64 rip;
-		USHORT fpcw;
-	};
+	} Position;
+
+	// MAX position use for computation
+	static const Position MAX = { (uint64_t)-1, (uint64_t)-1 };
```
Comment (commial) [inline TTD/TTD.hpp:45]:
```suggestion
	static const Position POSITION_MAX = { (uint64_t)-1, (uint64_t)-1 };
```

To avoid conflict with other global names

File: TTD/TTD.hpp [hunk 2]
```diff
@@ -270,7 +277,7 @@ namespace TTD {
 		void* unk9;
 		unsigned __int64(__fastcall* GetThreadCount)(TTD_Replay_ReplayEngine* self);
 		//	const struct TTD::Replay::ThreadInfo* (__fastcall* _GetThreadList_ReplayEngine_Replay_TTD__UEBAPEBUThreadInfo_23_XZ)(TTD::Replay::ReplayEngine* __hidden this);
```
Comment (commial) [inline TTD/TTD.hpp:279]:
```suggestion
```
No more needed

File: TTD/TTD.hpp [hunk 3]
```diff
@@ -437,7 +489,84 @@ namespace TTD {
...
+		 * \return	pointer to the first event
+		 */
+		const TTD_Replay_ModuleLoadedEvent* GetModuleLoadedEventList();
+
+		/*!
+		 * \brief	More c++ interface using vector
+		 *			list of module loaded event
+		 * \return	list of TTD_Replay_ModuleLoadedEvent
+		 */
+		const std::vector<TTD_Replay_ModuleLoadedEvent> GetModuleLoadedEvent();
```
Comment (commial) [inline TTD/TTD.hpp:512]:
```suggestion
		const std::vector<TTD_Replay_ModuleLoadedEvent> GetModuleLoadedEvents();
```
Should we add a `s` to all these interfaces, to be more meaningfull?

File: TTD/TTD.hpp [hunk 4]
```diff
@@ -437,7 +489,84 @@ namespace TTD {
...
+		 * \warning	use GetModuleUnloadedEventCount to know the number of event
+		 * \return	pointer to the first event
+		 */
+		const TTD_Replay_ModuleUnloadedEvent* GetModuleUnloadedEventList();
+
+		/*!
+		 * \brief	More c++ interface using vector
+		 * \return	vector of TTD_Replay_ModuleUnloadedEvent
+		 */
+		const std::vector<TTD_Replay_ModuleUnloadedEvent> GetModuleUnloadedEvent();
```
Comment (commial) [inline TTD/TTD.hpp:531]: Same remark for `s` suffix

File: TTD/TTD.hpp [hunk 5]
```diff
@@ -437,7 +489,84 @@ namespace TTD {
...
+		 * \warning	use GetThreadCreatedEventCount to know the number of event
+		 * \return	pointer to the first event
+		 */
+		const TTD_Replay_ThreadCreatedEvent* GetThreadCreatedEventList();
+
+		/*!
+		 * \brief	More c++ interface using vector
+		 * \return	vector of TTD_Replay_ThreadCreatedEvent
+		 */
+		const std::vector<TTD_Replay_ThreadCreatedEvent> GetThreadCreatedEvent();
```
Comment (commial) [inline TTD/TTD.hpp:550]: Same remark for `s` suffix

File: TTD/TTD.hpp [hunk 6]
```diff
@@ -437,7 +489,84 @@ namespace TTD {
...
+		 * \warning	use GetThreadTerminatedEventCount to know the number of event
+		 * \return	pointer to the first event
+		 */
+		const TTD_Replay_ThreadTerminatedEvent* GetThreadTerminatedEventList();
+
+		/*!
+		 * \brief	More c++ interface using vector
+		 * \return	vector of TTD_Replay_ThreadCreatedEvent
+		 */
+		const std::vector<TTD_Replay_ThreadTerminatedEvent> GetThreadTerminatedEvent();
```
Comment (commial) [inline TTD/TTD.hpp:569]: Same remark for `s` suffix

File: example_api/main.cpp [hunk 7]
```diff
@@ -30,7 +30,7 @@ int main()
 	int result;
 
 	std::cout << "Openning the trace\n";
-	result = ttdengine.Initialize(L"D:\\traces\\demo.run");
+	result = ttdengine.Initialize(L"C:\\Users\\sylvain\\Documents\\ttddbg_test_multithread02.run");
```
Comment (commial) [inline example_api/main.cpp:33]:
```suggestion
	result = ttdengine.Initialize(L"D:\\traces\\demo.run");
```

File: example_api/main.cpp [hunk 8]
```diff
@@ -71,7 +71,7 @@ int main()
 	DumpHex(ttdcursor.GetThreadInfo(), 0x20);
 
 	std::cout << "\nSet position to F3:0\n";
-	ttdcursor.SetPosition(0xF3, 0);
+	//ttdcursor.SetPosition(0xF3, 0);
```
Comment (commial) [inline example_api/main.cpp:74]:
```suggestion
	ttdcursor.SetPosition(0xF3, 0);
```

File: example_api/main.cpp [hunk 9]
```diff
@@ -80,11 +80,11 @@ int main()
 	DumpHex(ttdcursor.GetThreadInfo(), 0x20);
 
 	std::cout << "\nContext:\n";
-	TTD::TTD_Replay_RegisterContext* ctxt = ttdcursor.GetCrossPlatformContext();
+	auto ctxt = ttdcursor.GetContextx86();
```
Comment (commial) [inline example_api/main.cpp:83]:
```suggestion
        // Use  ttdcursor.GetContextx86() for x86 context
        auto ctxt = ttdcursor.GetCrossPlatformContext();
```
Is it OK for you tu use comment for x86 illustrations for now?

File: example_api/main.cpp [hunk 10]
```diff
@@ -80,11 +80,11 @@ int main()
...
 	DumpHex(ctxt, 0xA70);
-	printf("RCX: %llx\n", ctxt->rcx);
+	printf("RCX: %llx\n", ctxt->Ecx);
```
Comment (commial) [inline example_api/main.cpp:85]: As a result, can we keep the x64 example?
The x86 could be highlighted (in a future PR) by docs or another example

File: example_api/main.cpp [hunk 11]
```diff
@@ -159,4 +159,48 @@ int main()
...
+	while (itModuleLoaded != moduleLoaded.end() || itModuleUnloaded != moduleUnloaded.end() || itThreadCreate != threadCreate.end() || itThreadTerminate != threadTerminate.end())
+	{
+		TTD::Position moduleLoadedPosition = (itModuleLoaded == moduleLoaded.end()) ? TTD::MAX : itModuleLoaded->pos;
+		TTD::Position moduleUnloadedPosition = (itModuleUnloaded == moduleUnloaded.end()) ? TTD::MAX : itModuleUnloaded->pos;
+		TTD::Position threadCreatePosition = (itThreadCreate == threadCreate.end()) ? TTD::MAX : itThreadCreate->pos;
+		TTD::Position threadTerminatePosition = (itThreadTerminate == threadTerminate.end()) ? TTD::MAX : itThreadTerminate->pos;
+
+ 		if (moduleLoadedPosition < moduleUnloadedPosition && moduleLoadedPosition < threadCreatePosition && moduleLoadedPosition < threadTerminatePosition)
+		{
+			std::wcout << "Module Loaded at " << std::hex << itModuleLoaded->pos.Major << ":" << std::hex << itModuleLoaded->pos.Minor << " : " << std::wstring(itModuleLoaded->info->path) << std::endl;
```
Comment (commial) [inline example_api/main.cpp:184]: Maybe we should add a position display function to avoid such code

File: example_api/main.cpp [hunk 12]
```diff
@@ -159,4 +159,48 @@ int main()
...
+			itModuleLoaded++;
+		}
+
+		if (moduleUnloadedPosition < moduleLoadedPosition && moduleUnloadedPosition < threadCreatePosition && moduleUnloadedPosition < threadTerminatePosition)
+		{
+			std::wcout << "Module Loaded at " << std::hex << itModuleUnloaded->pos.Major << ":" << std::hex << itModuleUnloaded->pos.Minor << " : " << std::wstring(itModuleUnloaded->info->path) << std::endl;
```
Comment (commial) [inline example_api/main.cpp:190]:
```suggestion
			std::wcout << "Module Unloaded at " << std::hex << itModuleUnloaded->pos.Major << ":" << std::hex << itModuleUnloaded->pos.Minor << " : " << std::wstring(itModuleUnloaded->info->path) << std::endl;
```

File: TTD/TTD.hpp [hunk 13]
```diff
@@ -408,8 +408,60 @@ namespace TTD {
...
+		 * \brief	Get thread context, for the current one, for x86 target arch
+		 * \return	x86 thread context
+		 */
+		inline PWOW64_CONTEXT GetContextx86()
+		{
+			return static_cast<PWOW64_CONTEXT>(GetCrossPlatformContext());
+		}
+
+		/*!
+		 * \brief	Get thread context for x86_64 target arch
```
Comment (commial) [inline TTD/TTD.hpp:437]:
```suggestion
		 * \brief	Get thread context for x86 target arch
```

File: python-bindings/module.cpp [hunk 14]
```diff
@@ -77,6 +42,51 @@ PYBIND11_MODULE(pyTTD, m) {
...
+        }, nullptr)
+        .def_property("last_major", [](TTD::TTD_Replay_ActiveThreadInfo& self) {
+            return self.last.Major;
+        }, nullptr)
+        .def_property("last_minor", [](TTD::TTD_Replay_ActiveThreadInfo& self) {
+            return self.last.Minor;
+        }, nullptr);
+
+    py::class_<CONTEXT, std::unique_ptr<CONTEXT, py::nodelete>>(m, "Context")
+        .def_readonly("SegCs", &CONTEXT::SegCs)
```
Comment (commial) [inline python-bindings/module.cpp:63]: *cosmetics* I rather prefer (and found it more Pythonic) to have attribute in snake_case, starting with a lower letter

File: python-bindings/python-bindings.vcxproj [hunk 15]
```diff
@@ -124,7 +124,7 @@
       <SDLCheck>true</SDLCheck>
       <PreprocessorDefinitions>_DEBUG;_CONSOLE;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <ConformanceMode>true</ConformanceMode>
-      <AdditionalIncludeDirectories>$(SolutionDir)\TTD;C:\Python310\Include;C:\Python310\lib\site-packages\pybind11\include</AdditionalIncludeDirectories>
+      <AdditionalIncludeDirectories>$(SolutionDir)\TTD;C:\Program Files\Python38\include;C:\Program Files\Python38\Lib\site-packages\pybind11\include;C:\Python310\Include;C:\Python310\lib\site-packages\pybind11\include</AdditionalIncludeDirectories>
```
Comment (commial) [inline python-bindings/python-bindings.vcxproj:127]: Unnecessary changes
(And will disappear with #6 )

**Comment by commial** [comment]
Thanks for the PR!

Also, nice catching for the platform context.
Apart from the comment, LGTM

I'm opening #8 to track futures changes

File: example_api/main.cpp (hunk 11 above)
Comment (citronneur) [reply example_api/main.cpp:185]: Yes I'm afraid to write this code but as the two object are different, i didn't find a better way to do it...

File: python-bindings/module.cpp (hunk 14 above)
Comment (citronneur) [reply python-bindings/module.cpp:63]: OK I will rewrite this part (totally agree with the remark)
